File: packages/go/internal/seguranca/ataques/ataque_cobertura_dependencia.go
package ataques

import (
	"fmt"

	"criptografia/internal/seguranca"
)

// OpcoesCoberturaDependencia configura o AtaqueCoberturaDependencia.
type OpcoesCoberturaDependencia struct {
	TamanhoBloco       int
	PerturbacoesPorPar int
}

// AtaqueCoberturaDependencia verifica, byte a byte, se toda posição de SAÍDA
// depende de toda posição de ENTRADA da chave.
type AtaqueCoberturaDependencia struct {
	tamanhoBloco       int
	perturbacoesPorPar int
}

// NovoAtaqueCoberturaDependencia cria o ataque. Sem argumento, usa
// tamanhoBloco = 32 e perturbacoesPorPar = 5.
func NovoAtaqueCoberturaDependencia(opcoes ...OpcoesCoberturaDependencia) *AtaqueCoberturaDependencia {
	o := OpcoesCoberturaDependencia{TamanhoBloco: 32, PerturbacoesPorPar: 5}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueCoberturaDependencia{tamanhoBloco: o.TamanhoBloco, perturbacoesPorPar: o.PerturbacoesPorPar}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueCoberturaDependencia) Nome() string {
	return "Cobertura de dependência (matriz entrada x saída)"
}

// Executar roda o ataque.
func (a *AtaqueCoberturaDependencia) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	tamanho := a.tamanhoBloco
	chaveBase := make([]byte, tamanho)
	iv := seguranca.RandomBytes(alvo.TamanhoIv())

	ksBase := alvo.GerarKeystreamBruto(chaveBase, iv, "enc", tamanho)

	paresIndependentes := []string{}

	for posEntrada := 0; posEntrada < tamanho; posEntrada++ {
		afetouAlgumaSaida := make([]bool, tamanho)

		for p := 0; p < a.perturbacoesPorPar; p++ {
			chaveTeste := append([]byte(nil), chaveBase...)
			chaveTeste[posEntrada] = byte(seguranca.RandomInt(1, 255))
			ks := alvo.GerarKeystreamBruto(chaveTeste, iv, "enc", tamanho)

			for posSaida := 0; posSaida < tamanho; posSaida++ {
				if ks[posSaida] != ksBase[posSaida] {
					afetouAlgumaSaida[posSaida] = true
				}
			}
		}

		for posSaida := 0; posSaida < tamanho; posSaida++ {
			if !afetouAlgumaSaida[posSaida] {
				paresIndependentes = append(paresIndependentes, fmt.Sprintf("entrada[%d] -> saída[%d]", posEntrada, posSaida))
			}
		}
	}

	if len(paresIndependentes) > 0 {
		limite := min(20, len(paresIndependentes))
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: true,
			Severidade: seguranca.SevMedia,
			Detalhes: fmt.Sprintf(
				"%d par(es) sem dependência detectável em %d tentativas cada",
				len(paresIndependentes), a.perturbacoesPorPar,
			),
			Dados: map[string]any{"pares": paresIndependentes[:limite]},
		}, nil
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: false,
		Severidade: seguranca.SevInfo,
		Detalhes:   fmt.Sprintf("Todos os %d pares (entrada, saída) mostraram dependência", tamanho*tamanho),
	}, nil
}

File: packages/rust/src/seguranca/ataques/ataque_cobertura_dependencia.rs
//! Verifica, byte a byte, se toda posição de SAÍDA depende de toda posição
//! de ENTRADA (mudando 1 byte da chave, com várias perturbações diferentes
//! pra evitar falso-positivo por coincidência de valor). Uma dependência
//! ausente indica que a mistura não se propagou completamente - um atacante
//! poderia isolar e atacar aquele par de posições separadamente do resto.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::{random_bytes, random_int};

pub struct AtaqueCoberturaDependencia {
    pub tamanho_bloco: usize,
    pub perturbacoes_por_par: usize,
}

impl AtaqueCoberturaDependencia {
    pub fn new() -> Self {
        AtaqueCoberturaDependencia {
            tamanho_bloco: 32,
            perturbacoes_por_par: 5,
        }
    }
}

impl Default for AtaqueCoberturaDependencia {
    fn default() -> Self {
        Self::new()
    }
}

impl Ataque for AtaqueCoberturaDependencia {
    fn nome(&self) -> String {
        "Cobertura de dependência (matriz entrada x saída)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let tamanho = self.tamanho_bloco;
        let chave_base = vec![0u8; tamanho];
        let iv = random_bytes(alvo.tamanho_iv());

        let ks_base = alvo.gerar_keystream_bruto(&chave_base, &iv, "enc", tamanho);

        if chave_base.len() != tamanho {
            // A chave precisa ter o mesmo tamanho do bloco pra esse teste
            // isolar 1 posição de cada vez sem wraparound. Se a cifra usa
            // chave de outro tamanho, pula esse ataque.
            return Err(ErroAtaque::Skip(
                "Teste requer chave do mesmo tamanho do bloco.".to_string(),
            ));
        }

        let mut pares_independentes: Vec<String> = Vec::new();

        for pos_entrada in 0..tamanho {
            let mut afetou_alguma_saida = vec![false; tamanho];

            for _ in 0..self.perturbacoes_por_par {
                let mut chave_teste = chave_base.clone();
                chave_teste[pos_entrada] = random_int(1, 255) as u8;
                let ks = alvo.gerar_keystream_bruto(&chave_teste, &iv, "enc", tamanho);

                for pos_saida in 0..tamanho {
                    if ks[pos_saida] != ks_base[pos_saida] {
                        afetou_alguma_saida[pos_saida] = true;
                    }
                }
            }

            for (pos_saida, &afetou) in afetou_alguma_saida.iter().enumerate() {
                if !afetou {
                    pares_independentes.push(format!(
                        "entrada[{}] -> saída[{}]",
                        pos_entrada, pos_saida
                    ));
                }
            }
        }

        if !pares_independentes.is_empty() {
            return Ok(ResultadoAtaque::new(
                &self.nome(),
                true,
                Severidade::Media,
                &format!(
                    "{} par(es) sem dependência detectável em {} tentativas cada",
                    pares_independentes.len(),
                    self.perturbacoes_por_par
                ),
            ));
        }

        Ok(ResultadoAtaque::new(
            &self.nome(),
            false,
            Severidade::Info,
            &format!(
                "Todos os {} pares (entrada, saída) mostraram dependência",
                tamanho * tamanho
            ),
        ))
    }
}
